### ISSUE TITLE
Improving Reloaded's Reputation through Improving User Experience

### ISSUE THREAD

**Issue by Sewer56** [Reloaded-Project/Reloaded-II#47]
# Background

I have always felt as if Reloaded has always had a bad reputation among mod loaders due to users' inability to set it up properly (even though it is not difficult!). I often witness various threads for mods and guides showered with identical copycat questions as opposed to positive comments, which in my opinion is a great deterrent from the eyes of a potential new user.

Normally, end users take things to work out of the box for granted; i.e.
- Mod launcher boots on first try.
- Magically detects the game.
- User adds mods. 
- Presses `Play` and it runs. 

For technical reasons, this is a bit difficult due to Reloaded's situation of being a universal loader.
You can't possibly know what the end user will use the loader with, and because it runs on .NET Core (which does not ship with Windows), users are required to download Core manually. Very often they forget to do this and later complain in places where they might sometimes not get help, which litters both guides, community servers and release pages for Reloaded.

Download/Install issues (in 2020) mostly stem from illiterate or impatient users not willing to read, as the requirements are very clearly outlined ([sometimes in the same guides people complain in!](https://gamebanana.com/tuts/13379)), however you must also consider users without a good level of English literacy. We should auto install the runtimes in the future if possible.

Another popular misconception among new users is believing that they only require x64 version of Core. While the Github version outlines this, this isn't as easy to outline on GameBanana (and some users just skim the text), hence there are also often duplicate complaints when `Kernel32AddressDumper` fails (BasicDllInjector failure).

For many users, mod discoverability is also an issue. This is unfortunately due to the fact that I have not always bothered to look at things from a PR perspective despite creating this loader (I should be setting a good example). For me, it was always about making good things, and not taking time out to market them; however, what is the point if nobody knows about the good stuff? 
There are plenty of miscellaneous changes that should be made towards improving mod marketability, many which are outlined below. 

# Measures
✔ Indicates Complete and Available in Current Development Version
❓ Indicates ToDo or to be Finalized.
❌ Indicates Not Yet Available

## ✔ Automatic Mod Loader Dependency Checking
People are bad at reading.
See issue #43 for extra details.

## ✔ Automatic Mod Dependency Downloads
Already implemented.

## ✔ More Descriptive Error Messages
There are certain error messages which are triggered by common user errors or loader bugs which perhaps may be better represented. For example...

- When .NET Core x86 is Missing (DLL Injector Failure):
![](https://user-images.githubusercontent.com/40992017/84752187-0bb1ce00-af83-11ea-99d2-d8542a4e72c8.png)

- When Loader Config gets Corrupted due to Unknown Cause (#34)
![](https://user-images.githubusercontent.com/68884141/88612520-4148dd00-d040-11ea-8af4-31dcb45d3fc9.png)

These exceptions should probably state the most likely user error and how to fix it.

## ❌ User Focused Documentation
Try to target user facing documentation more towards end users in wording as opposed to power users.
e.g. [Injection Methods](https://github.com/Reloaded-Project/Reloaded-II/blob/master/Docs/InjectionMethods.md) should probably have a better name and use more simple wording.

It might also be useful to add information about using Reloaded with e.g. Steam into the documentation.

## ❓ Better Mod Marketability
Here are some areas to work on, with regards to making mods (at least my own) more marketable.

- Use more `Epic` keywords in more major mods.
   - e.g. Revamp, Remaster, Overhaul
   - Do not use technical terms e.g. Hook

- Tweet. Social Media exists for a reason.
  - Use custom banner for each individual release (Gamebanana too).
  - YouTube video where appropriate.

- Create custom banners for mods.
  - I was initially lazy but it helps for categorisation. Template banners are really really bad.
  - Many mod submissions need more images.

- User targeted description/readmes.
  - Current ones are targeted for developers which are a no-no.
  - For example: [Heroes Controller Hook](https://gamebanana.com/gamefiles/9741?post=8612950) is very barebones. It's not clear enough for users to use Post Process mod to flip triggers etc.
  - Need more screenshots?

- There exist many **important** game specific and non-game specific mods.
   - How to we make them discoverable?
   - Universal mods such as [Universal File Redirector](https://github.com/Reloaded-Project/reloaded.universal.redirector) especially come to mind.

**Comment by Arggg** [comment]
Better Mod Marketability is the only topic I can really comment on. Asking the user if they're ok with downloading the dependencies or closing should be fairly clear. In terms of this though, I have some thoughts.

1. Develop a formatting style for reloaded mods and indicate to developers that you wish it'd be used. 
As an example: 
GAME | MOD LOADER | TOOL NAME | TOOL TYPE
Universal | Reloaded 2 | controller hook | inputs
Or something similar - we just want something that everyone can follow that will make it clearer for the user
2. If you're going to make custom banners for a bunch of mods, does RII already let you have custom banners, do you intend to impliment this, or is this just for game banana?

**Comment by Sewer56** [comment]
Custom Banner Example:
![image](https://user-images.githubusercontent.com/6697380/90807875-0a9c7480-e317-11ea-8fe2-ece7b3881fa6.png)

But another custom one with widescreen for social media/GB releases.

**Comment by Arggg** [comment]
Also a default one w

**Comment by Sewer56** [comment]
Current default one is the Reloaded logo, which is a bit weird honestly: 
![image](https://user-images.githubusercontent.com/6697380/90852678-06507580-e370-11ea-9631-7f9fdf63b037.png)

Probably want some sort of background since there's an outline the banner.
Maybe something similar to my profile avatar (on Steam) 
![image](https://user-images.githubusercontent.com/6697380/90852889-a0b0b900-e370-11ea-89cc-1acf51db5783.png)

**Comment by Arggg** [comment]
I've been thinking about changing that part of the ui a lot tbh

**Comment by Albeoris** [comment]
For developing mods for Unity, the most convenient tool now is BepInEx.
Therefore, I consider Reloaded primarily as a tool for working with native code.
There are few such mods, most often - one per game. For such scenarios, I would like to have an out-of-the-box solution for the end user that allows you to build an installer for one specific mod for one specific game.

**Comment by AlmightyLks** [comment]
Feedback regarding improving User Experience:  
I've found this project a few minutes ago through the github page at https://reloaded-project.github.io/Reloaded-II/ .  
On there I could not, for the death of me, find a guidance on how this project / program is to be used.  
I assumed some kind of setup as bepinex - as albeoris mentioned above - but I wasnt sure.  
The overview on the left hand side doesnt signal anything like that.  
A search for "download" did it, leading to github (besides gamebanana).  
On GitHub, the readme is minimalistic, directing to the github page. Fair.  
The release however is flooded with 7 different assets. Nowhere does it really tell me what these are and what to choose.  
Setup.exe didn't work due to an error (#190 ). So I opted for a manual install via the Release.zip.  

Now I've tried setting up a Proof-Of-Concept Mod to try out this project, and trying to build the bare-metal mod template in VS 
 tells me that the "RELOADEDIIMODS" environment variable is missing. I assume that would come with the normal setup.exe? Idk. Looking through the github page, I cant find anything about any env vars or "RELOADEDIIMODS". Making a search for that on github also doesnt return any known issues.  
Turns out the program folder has that mod-folder, and I _think_ I am supposed to use that directory for the system env var

Welp

**Comment by Sewer56** [comment]
While I'm not really sure what happened in the first case (that DLL is part of the .NET runtime, it shouldn't be missing), I can speak for the second one.

This environment variable is set by the launcher every time it boots as part of the boot procedure, and points to your own `Mods` folder (which is also possible to move with manual config edit).  The idea is that you can clone any project on GitHub and hit 'build' without having to adjust paths etc. and have the mod automatically appear in your mod list.

I'm guessing what happened is you had an IDE opened before opening the launcher for the first time and thus the environment variable didn't register for the IDE. Rebooting IDE should fix it.

**Comment by AlmightyLks** [comment]
You're right! 
Interesting, didnt take that into consideration
But there was no way for me to know when it sets that var tbf 😅

**Comment by RandomAnimeGamer** [comment]
I know this is super-old, but this was the only post I could find via a search that linked directly to the Github of Reloaded. I feel there are several things that can be done to better the user experience.

First, I want to clarify dependencies. Mod dependencies are not always clearly laid out by the authors (and sometimes downright missing entirely), but they may have listed them in the mod page metadata. Expecting users to always know every dependency is a tall order, in my opinion, given how differently developers will actually document these things. One example is Reloaded itself, not having great documentation despite being functional in those areas.

I'd also like to describe the experience installing mods when you have more than one game. I had previously loaded mods via Persona 4 Golden on PC, which was relatively easy. I learned some mods for Dragon Ball Sparking Zero required Reloaded II, which I wasn't particularly happy with given I'd been used to using Unverum (which I prefer heavily over Reloaded II) for DBSZ.

These mod pages don't have any instructions on installing besides to "follow guides on installing Reloaded II online". So I use my pre-existing Reloaded II install to add Sparking Zero. Easy enough. Then it comes to installing mods. Unverum has a 1-click install via Gamebanana and also supports drag and drop, so I opt to drag and drop mod archives onto Reloaded II, with Sparking Zero in-context. Nothing happens. I drag and drop again. Nothing happens. I try looking through menus, and found it installed over Persona 4 Golden.

This is my first gripe. If I've selected the game in question to be in-context, the app should recognize this when attempting to install mods. Furthermore, it had two entries of the mod now, when there should never be more than one of the same exact mod data for the same exact game.

I remove these two mods, and try again with the same steps as above, with the same behavior. I try to drag the menu entries on P4G to the Sparking Zero icon, nothing happens. I look around the file system to find that the mods are all in their own one folder, not separated by game, meaning I couldn't drag it from one folder to another. I understand some mods _may_ be shared between games, but this isn't a common case and it could easily be mitigated by having a "Shared" folder for mods that apply to that scenario.

I look through the UI and cannot find for the life of me how to move these mods from Persona 4 Golden to Sparking Zero. My first idea was "manage mods", but by this point I wasn't paying 100% attention to every detail, and got thoroughly confused at that menu. By default it has a mod already selected on the left hand side, but not on the list on the right hand side.

I even try to find a "default game" in the settings, since it seemed to only install mods for Persona 4 Golden. I find the GUI of Reloaded II very unintuitive outside of the sidebar - the only piece of UI that makes sense to me.

Eventually I find out that I'm supposed to go to Manage Mods, go to the mod I installed, make sure Persona 4 Golden shows a "-" sign and Sparking Zero shows a "+" sign (rather than a checkbox, something every other app uses?) to move them.

Now imagine someone has more than just a few mods, but a huge list of mods. It doesn't make sense to have such a large menu to move mods, something that could easily be accomplished via context menu either way.

This is also compounded with a button on each game that says "Configure Mods" not to be confused with the sidebar button that says "Manage Mods".

What was supposed to be an easy task of a drag-and-drop became an odyssey of mishaps and confusion led by an unintuitive GUI, in my opinion.

The case of Reloaded's reputation can entirely be attributed to its poor user experience, spearheaded by its unclear GUI. Pretending as if peoples' problems with Reloaded entirely are summed up by "the users are stupid, people can't read, and we don't market it well" will only make your reputation worse, as now I'm even less inclined to use Reloaded II for anything else.

I'm being harsh about this because my full time job is programming and all of what you said in your initial post would be absolutely unacceptable at any professional capacity.

If you want any possible way to improve, if you're absolutely unwilling to listen to user feedback and be as cynical as possible, try looking at mod loaders people actually _enjoy_ using. Unverum is the most widely used for the games it does support for a very good reason. SA Mod Loader is the most widely used for Sonic Adventure games for a very good reason. If the issue was 100% "users can't read" then why do these other mod loaders consistently have better reputations than Reloaded? Look inward and you'll see.

**Comment by Sewer56** [comment]
Hi, thanks for the feedback.
I'm closing the issue as it's long no longer relevant, however I should nonetheless address the post.

> First, I want to clarify dependencies. Mod dependencies are not always clearly laid out by the authors (and sometimes downright missing entirely), 

For around half the games, mods only need to set dependency on a single mod, e.g. `Persona Essentials`, some exclusions apply when targeting a modular extension.

In the future I hope to show warnings if it is believed a mod does not specify a dependency it should via 'Diagnostics'; think something like IDE error window. Not in Reloaded-II, but in its successor.

> Then it comes to installing mods. Unverum has a 1-click install via Gamebanana and also supports drag and drop

1-click install via GameBanana requires asking the site owner to enable it for a specific game page. If it is not present for a valid Reloaded mod, you need to ask `tom` from GameBanana to enable it for that game.

Drag & Drop is also supported, though some odd controls swallow the drop for under the hood reasons, I haven't gotten around to fixing that.

>  I opt to drag and drop mod archives onto Reloaded II, with Sparking Zero in-context. Nothing happens. I drag and drop again. Nothing happens. I try looking through menus, and found it installed over Persona 4 Golden.

Mods expose what games they support through a field in the `ModConfig.json` called `SupportedAppId`. Alternatively if they set the universal mod flag, they may show in all games.

There is no context to speak of, mods all live in a single bucket (mods folder) and opt in into showing themselves for a game. The ID assigned to a game, by default is the lowercase name of the game's executable. Provided you have not renamed it, adding a mod should make it automatically show in the right game no matter how you added it.

A mod author is also unlikely to misconfigure this, because the mod would not show on their own end if they renamed the binary.

In the successor, universal mods will be able to specify 'groups' (tags) of games they should show up in, and mods will opt into being shown in those groups, but for now universal/multi-game mods show globally.

> My first idea was "manage mods", 
> make sure Persona 4 Golden shows a "-" sign and Sparking Zero shows a "+" sign (rather than a checkbox, something every other app uses?) to move them.

Yeah that menu is horrible, you have my condolences there. It's pretty much the only unchanged menu since initial release in 2019 as it barely sees any use.

> all of what you said in your initial post would be absolutely unacceptable at any professional capacity.

I have to respectfully disagree. Even if the wording may be far from perfect, the initial post was purely factual. (I was also a teenager at the time, please cut me some slack)

The intent was to say that installing the launcher was too difficult, for people whose first language was not English, and for people with little patience who wish to get started quickly.

In 2020, the situation was a bit different:

- There was no installer, only `Release.zip`.
- Running the launcher (.NET Core 3.0) without the runtime installed led to a completely silent crash.
- The user was manually required to install the runtime.

The runtime installation requirement was noted in the following places:

- On the Release file, on GameBanana.
- Directly above the release file, on GameBanana (so twice)
- On the repository's main readme.
- In GitHub Releases, directly above `Release.zip`
- On the main landing page for the project's wiki.
- In dedicated installation guides, such as the one hyperlinked in the original post.

On every page with a download link it was present twice, in the main body of the text, and directly above or on the download button.

Unfortunately that wasn't good enough, as a lot of questions about nothing happening when they launch the .exe kept appearing, including dedicated install guides, which led to me opening this issue at the time.

It was particularly a problem because people would ask for help in unusual places (for example Reddit), where they were unlikely to be helped as nobody knowledgeable was watching; so I've tried my best at the time to make sure that wasn't the case.

And later down the road I was able to make an automated installer, by reading the .NET Runtime specifications and the CI dotnet-install script and writing a runtime crawler to detect installed versions and then auto downloader.

> Unverum and SA Mod Loader are the most widely used for the games they support for a very good reason.

I'm sorry. I'm physically unable to compete with these projects. 

Unverum is much smaller in scope (mod manager only) and development is partially funded by commissions. It doesn't need a dedicated mod loader or core game plugins as that's served by UE4SS.

SA Mod Loader, HedgeModManager etc. are maintained by a team of active contributors, both projects have 10 or more contributors who contributed more than all contributions I have received combined. The scope is Mod Manager + Mod Loader.

Over here, it's always been a solo job. I singlehandedly always had to maintain the Mod Manager, Mod Loader, (Some) Core Game Mods and Core Libraries used by game mods, including Hooking, Sig Scans, Caching etc. I also have to provide tech support in a timely manner, usually within 5 minutes if I'm awake.

My scope is larger than the other projects and I (unfortunately) always have had to tackle it alone; and it's (always) been a struggle. In any case, I originally built the project to give some very niche games the opportunity to be modded in more advanced ways, I've never expected people would start using it in more mainstream games; unfortunately, admittedly parts of it aren't the best equipped to handle modern games.

I've been trying to build a better successor to tackle the issues, but it's been taking time, a lot of time. I've been writing code for 70+ hours a week (including full time job) for around the past 88/94 weeks, and it's still not good enough. I do it for the people, not myself, expecting nothing in return. It may take another 2 years until I can get to an alpha. If I don't end up killing myself by then.

So I apologize if today's Reloaded-II does not live up to your expectations. I'm simply unable to do a better job at the moment with the limited resources I have at hand.

**Comment by RandomAnimeGamer** [comment]
> So I apologize if today's Reloaded-II does not live up to your expectations. I'm simply unable to do a better job at the moment with the limited resources I have at hand.

Then let me ask you this - are you open to contributors? It sounds like there's a lot of work to be done, and if you're the literal only person working, there must be another out there (including myself) willing to help.

Unless you're absolutely turning down the help, then that's another story.

**Comment by Sewer56** [comment]
I'm always open to contributions, provided they follow the general spirit of the project (for example, don't drop files into game folder unless absolutely necessary). Largest contributed feature I can think of has been persisting mod order across reboots, for those who don't like their mods auto-sorted by enabled then disabled.

Examples of things up for grabs (some simple things):

- https://github.com/Reloaded-Project/Reloaded-II/issues/479 
    Wine: Throw a friendlier exception message asking users them to give FlatPak Protontricks, FileSystem permission. Ideally, check if it exists before starting the install at all by trying to drop a dummy file.
- https://github.com/Reloaded-Project/Reloaded-II/issues/481#issuecomment-2460439602
    Filter away colons `-` and invalid characters in game names.
- https://github.com/Reloaded-Project/Reloaded-II/issues/477
    Sanitizing the default install location of Reloaded against special characters and OneDrive.
- https://github.com/Reloaded-Project/Reloaded-II/issues/269
    A prompt for deleting mods.
- https://github.com/Reloaded-Project/Reloaded-II/issues/150
    Deploy a 1 click protocol handler on Wine / Proton (same way a `.desktop` file is deployed by installer for shortcuts)
- https://github.com/Reloaded-Project/Reloaded-II/issues/492
    A message informing the user how to uninstall if they used `Deploy ASI Loader` to perma-install Reloaded to their game and then removed the Reloaded folder. (Basically print DLL location in `Reloaded.Mod.Bootstrapper` telling them to delete file if loader removal is intended.)

It's mostly smaller scale issues. Although I'm not actively adding any new features to Reloaded-II, as I'm working on [the successor](https://github.com/Reloaded-Project/Reloaded-III) (right now it's just planning docs, a lot of docs), I do fix any major bugs that pop up. So what's generally left are the smaller issues.
